File: solutions/0/0/2/1/p0021_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 21: Amicable Numbers [Level 1]. """
from __future__ import annotations

import functools

from solver.runners import runner


@runner.main
def solve(*args: str) -> str:
    """Sum every amicable number in [2, max_num].

    For each x let y = d(x), the sum of proper divisors; x is amicable when
    y != x and d(y) == x. d(n) is computed by O(sqrt(n)) divisor pairing and
    memoised per call, giving an overall O(max_num * sqrt(max_num)) cost. The
    cache is built inside solve() so each benchmark run pays its full cost.
    """
    max_num = runner.parse_int(args[0])

    @functools.lru_cache(maxsize=None)
    def sum_factors(n: int) -> int:
        """Sum of the proper divisors of n, via square-root divisor pairing."""
        n_sqrt = int(n**0.5)
        return 1 + sum((i + n // i for i in range(2, n_sqrt + 1) if n % i == 0)) - (n_sqrt if n_sqrt**2 == n else 0)

    return str(sum((x for x in range(2, max_num + 1) if (y := sum_factors(x)) != x and sum_factors(y) == x)))


if __name__ == "__main__":
    raise SystemExit(solve())

File: solutions/public/p0021/p0021_s0.c
/* Solution to Euler Problem 21: Amicable Numbers. */
#include "runner.h"
#include <math.h>

static long long sum_factors(int n) {
    int n_sqrt = (int)sqrt((double)n);
    long long total = 1;
    for (int i = 2; i <= n_sqrt; i++) {
        if (n % i == 0) {
            total += i + n / i;
        }
    }
    if (n_sqrt * n_sqrt == n) {
        total -= n_sqrt;
    }
    return total;
}

const char *solve(int argc, char *argv[]) {
    static char _answer[32];
    int max_num = parse_int(argv[1]);

    long long result = 0;
    for (int x = 2; x <= max_num; x++) {
        long long y = sum_factors(x);
        if (y != x && sum_factors((int)y) == x) {
            result += x;
        }
    }
    { snprintf(_answer, sizeof _answer, "%lld", (long long)(result)); return _answer; }
}
